enhancements/proposals/node-autoreplacement.md

```diff
@@ -0,0 +1,114 @@
+# Node Autoreplacement
+
+## Summary
+
+Node Autoreplacement enhancement extends Scylla Operator with a functionality of automated replacement of nodes which lost their storage, which, at this point, renders them getting stuck on joining the cluster.
+
+This proposal suggests creating a module responsible for triggering a node replacement in such circumstances.
+As a side effect, introducing such functionality would allow for simplifying the existing `OrphanedPVController` by
+reducing its task to evicting orphaned PVs.
+
+## Motivation
+
+Currently, when a node looses its storage, it gets stuck on joining a cluster, since Scylla produces a following runtime error:
+```
+Startup failed: std::runtime_error (A node with address <ip> already exists, cancelling join. Use replace_address if you want to replace this node.)
+```
+
+Successfully restarting the node requires a manual intervention of a human operator by annotating the
+service corresponding to the affected node with a `scylla/replace` label. The main goal of this proposal is to improve
+Scylla Operator's stability in running Scylla independently of its administrators' interventions.
+
+### Goals
+
+- Successfully discovering and autoreplacing nodes which lost their storage without requiring user intervention.
+
+### Non-Goals
+
+- Removing orphaned PVs. This is left for `OrphanedPVController`.
+
+## Proposal
+
+Create an InitContainer with a dedicated Scylla Operator command responsible for checking whether a given node is
+1. **not** first in the cluster
+2. already exists in gossip
+
+In case both of these conditions are true, the autoreplace module triggers running Scylla with `replace-address-first-boot` flag.
+
+It is necessary to check whether a node is first in the cluster, as the first node in the cluster skips bootstrap, and therefore the flag can't be set.
+Checking for the IP already existing in gossip is necessary as it is not possible to replace a node that does not.
+Setting `replace-address-first-boot` results in a desired output, i.e. a node which lost its storage will be replaced during bootstrap.
+Otherwise, setting the flag is safe as Scylla ignores it after bootstrap.
+
+Introducing this enhancement additionally allows for removing replacement functionality from `OrphanedPVController`. The controller would be simplified and only be responsible for removing the orphaned PVs.
+
+### User Stories
+
+#### Story 1
+
+The user is running a ScyllaCluster and removes a PVC bound to one of Scylla nodes. The node gets automatically replaced. 
+
+#### Story 2
+
+The user is running a ScyllaCluster on AWS/GCP cluster with a local storage. One of the instances gets restarted, which results in erasing the local storage.
```

Comment:
I think restart keeps the local storage, suspend, hibernate and others don't

Reply:
Explicit restart of an instance with local SSD is technically not allowed, but I've done the following:
1. Deploy a GKE cluster with 2 nodes
2. Wait for it to deploy
3. ssh to one of the instances and run `sudo shutdown -h now`
4. Wait for it to get back up again
And it resulted with losing storage and the pod logging `Startup failed: std::runtime_error (A node with address 10.55.254.64 already exists, cancelling join. Use replace_address if you want to replace this node.)`

Reply:
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/InstanceStorage.html
https://cloud.google.com/compute/docs/disks/local-ssd#data_persistence